llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,138 @@ getDoubleGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
     return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
 }
 
+namespace {
+  struct KernelProperties {
+    KernelProperties() = default;
+
+    KernelProperties(const Function *F) {
+      if (const MDNode *MDN = F->getMetadata("sycl_used_aspects")) {
+        auto ExtractIntegerFromMDNodeOperand = [=](const MDNode *N,
+                                                   unsigned OpNo) -> auto {
+          Constant *C =
+              cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+          return C->getUniqueInteger().getSExtValue();
+        };
+
+        for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+          Aspects.insert(ExtractIntegerFromMDNodeOperand(MDN, I));
+        }
+
+      }
+    }
+
+    std::string getName(StringRef BaseName) const {
+      if (Aspects.empty())
+        return BaseName.str() + "-no-aspects";
+
+      std::string Ret = BaseName.str() + "-aspects";
+      for (int A : Aspects) {
+        Ret += "-" + std::to_string(A);
+      }
+      return Ret;
+    }
+
+    SmallSet<int, 4> Aspects;
+    // TODO: extend this further with reqd-sub-group-size, reqd-work-group-size,
+    // double-grf and other properties
+
+    static KernelProperties getTombstone() {
+      KernelProperties Ret;
+      Ret.IsTombstoneKey = true;
+      return Ret;
+    }
+
+    static KernelProperties getEmpty() {
+      KernelProperties Ret;
+      Ret.IsEmpty = true;
+      return Ret;
+    }
+
+  private:
+    // For DenseMap:
+    bool IsTombstoneKey = false;
+    bool IsEmpty = false;
+
+  public:
+    bool operator==(const KernelProperties &Other) const {
+      // Tombstone does not compare equal to any other item
+      if (IsTombstoneKey || Other.IsTombstoneKey)
+        return false;
+
+      if (Aspects.size() != Other.Aspects.size())
+        return false;
+
+      if (!llvm::all_of(Other.Aspects,
+                        [&](int Aspect) { return Aspects.contains(Aspect); }))
+        return false;
+
+      return IsEmpty == Other.IsEmpty;
+    }
+
+    unsigned hash() const {
+      llvm::hash_code AspectsHash =
+          llvm::hash_combine_range(Aspects.begin(), Aspects.end());
```

Comment:
NIT: this could be relatively heavy operation, result could be cached. Maybe add a TODO?

Reply:
Implemented a cache mechanism in 7d7fd45eead5c53ff1eb4f96a19740f8e145b66d